File: src/main/scala/org/soaringmeteo/LocationForecasts.scala
package org.soaringmeteo

import java.time.{LocalDate, OffsetDateTime}
import java.time.format.DateTimeFormatter

import io.circe.{Encoder, Json}
import org.soaringmeteo.GfsForecast.pressureLevels
import squants.energy.SpecificEnergy
import squants.motion.Pressure
import squants.radio.Irradiance
import squants.space.Length
import squants.thermal.Temperature

import scala.util.chaining._

/**
 * All the forecast data for one location and several days, with
 * high-level information extracted from the content of each GFS forecast
 * (e.g., thunderstorm risk per day)
 */
case class LocationForecasts(
  elevation: Length,
  dayForecasts: Seq[DayForecast]
)

case class DayForecast(
  date: LocalDate,
  hourForecasts: Seq[DetailedForecast],
  thunderstormRisk: Int /* Between 0 and 4, for the entire day (not just the forecast period. */
)

case class DetailedForecast(
  time: OffsetDateTime,
  boundaryLayerHeight: Length,
  boundaryLayerWind: Wind,
  cloudCover: CloudCover,
  atPressure: Map[Pressure, IsobaricVariables],
  mslet: Pressure,
  snowDepth: Length,
  surfaceTemperature: Temperature,
  surfaceDewPoint: Temperature,
  surfaceWind: Wind,
  totalRain: Length,
  convectiveRain: Length,
  latentHeatNetFlux: Irradiance,
  sensibleHeatNetFlux: Irradiance,
  cape: SpecificEnergy,
  cin: SpecificEnergy,
  downwardShortWaveRadiationFlux: Irradiance,
  isothermZero: Length
)

object LocationForecasts {

  def apply(location: Point, forecasts: Seq[GfsForecast]): LocationForecasts = {
    val isRelevantAtLocation = isRelevant(location)
    val forecastsWithRainPerPeriod: Seq[DetailedForecast] =
      forecasts.foldLeft((List.newBuilder[DetailedForecast], Option.empty[GfsForecast])) {
        case ((builder, maybePreviousForecast), forecast) =>
          val (totalRain, convectiveRain) =
            maybePreviousForecast match {
              case None =>
                // First forecast (+3h) is special, it contains the accumulated
                // rain since the forecast initialization time, which is equivalent
                // to the rain that fell during the forecast period
                (forecast.accumulatedRain, forecast.accumulatedConvectiveRain)
              case Some(previousForecast) =>
                (
                  forecast.accumulatedRain - previousForecast.accumulatedRain,
                  forecast.accumulatedConvectiveRain - previousForecast.accumulatedConvectiveRain
                )
            }

          builder += DetailedForecast(
            forecast.time,
            forecast.boundaryLayerHeight,
            forecast.boundaryLayerWind,
            forecast.cloudCover,
            forecast.atPressure,
            forecast.mslet,
            forecast.snowDepth,
            forecast.surfaceTemperature,
            forecast.surfaceTemperature * forecast.surfaceRelativeHumidity / 100,
            forecast.surfaceWind,
            totalRain,
            convectiveRain,
            forecast.latentHeatNetFlux,
            forecast.sensibleHeatNetFlux,
            forecast.cape,
            forecast.cin,
            forecast.downwardShortWaveRadiationFlux,
            forecast.isothermZero
          )
          (builder, Some(forecast))
      }._1.result()
    LocationForecasts(
      elevation = forecasts.head.elevation,
      dayForecasts =
        forecastsWithRainPerPeriod
          .filter(forecast => isRelevantAtLocation(forecast.time)) // Keep only forecasts during the day, and around noon
          .groupBy(_.time.toLocalDate)
          .filter { case (_, forecasts) => forecasts.nonEmpty }
          .toSeq
          .sortBy(_._1)
          .map { case (forecastDate, hourForecasts) =>
            DayForecast(
              forecastDate,
              hourForecasts.sortBy(_.time),
              if (hourForecasts.sizeIs == 3) thunderstormRisk(hourForecasts(0), hourForecasts(1), hourForecasts(2))
              else {
                // TODO Print warning?
                0
              }

            )
          }
    )
  }

  def isRelevant(location: Point): OffsetDateTime => Boolean = {
    // Transform longitude so that it goes from 0 to 360 instead of 180 to -180
    val normalizedLongitude = 180 - location.longitude
    // Width of each zone, in degrees
    val zoneWidthDegrees = 360 / Settings.numberOfForecastsPerDay
    // Width of each zone, in hours
    val zoneWidthHours   = Settings.gfsForecastTimeResolution
    // Noon time offset is 12 around prime meridian, 0 on the other side of the
    // earth, and 6 on the east and 21 on the west.
    // For example, a point with a longitude of 7 (e.g., Bulle) will have a normalized
    // longitude of 173. If we divide this number of degrees by the width of a zone,
    // we get its zone number, 4. Finally, we multiply this zone number by the number of
    // hours of a zone, we get the noon time for this longitude, 12.
    val noonHour =
      ((normalizedLongitude + (zoneWidthDegrees / 2.0)) % 360).doubleValue.round.toInt / zoneWidthDegrees * zoneWidthHours

    val allHours = (0 until 24 by Settings.gfsForecastTimeResolution).to(Set)

    val relevantHours: Set[Int] =
      (1 to Settings.relevantForecastPeriodsPerDay).foldLeft((allHours, Set.empty[Int])) {
        case ((hs, rhs), _) =>
          val rh = hs.minBy(h => math.min(noonHour + 24 - h, math.abs(h - noonHour)))
          (hs - rh, rhs + rh)
      }._2

    time => {
      relevantHours.contains(time.getHour)
    }
  }

  // FIXME Generalize to an arbitrary number of forecasts for the day
  def thunderstormRisk(
    forecastMorning: DetailedForecast,
    forecastNoon: DetailedForecast,
    forecastAfternoon: DetailedForecast
  ): Int = {
    // Blindly copied from src/makeGFSJs.pas. We seriously need to check the formulas again.
    val baseCAPE = (forecastNoon.cape.toGrays + forecastAfternoon.cape.toGrays) / 100
    val factorCAPE =
      if (baseCAPE >= 2) baseCAPE
      else baseCAPE - 4 * (2 - baseCAPE)

    val spreadAltiMorning   = spreadAlti(forecastMorning)
    val spreadAltiNoon      = spreadAlti(forecastNoon)
    val spreadAltiAfternoon = spreadAlti(forecastAfternoon)
    val factorSpread        = (spreadAltiMorning + spreadAltiNoon + spreadAltiAfternoon) / 3

    val factorSensibleHeat =
      (forecastMorning.sensibleHeatNetFlux.toWattsPerSquareMeter / 10).min(10)
      .pipe { rawValue =>
        if (rawValue >= 3) rawValue
        else rawValue - 3 * (3 - rawValue)
      }

    val factorConvection = (
      (
        forecastMorning.cloudCover.conv +
        forecastNoon.cloudCover.conv +
        forecastAfternoon.cloudCover.conv
      ) / 10 +
      forecastMorning.convectiveRain.toMillimeters +
      forecastNoon.convectiveRain.toMillimeters +
      forecastAfternoon.convectiveRain.toMillimeters
    ) / 2

    val factorG = factorCAPE + factorConvection + factorSensibleHeat - factorSpread

    if (factorG < -3) 0
    else if (factorG < 3) 1
    else if (factorG < 8) 2
    else if (factorG < 18) 3
    else 4
  }

  /** Average “spread” value for altitudes above 3000 m */
  def spreadAlti(forecast: DetailedForecast): Double = {
    spreadAndTds(forecast)
      .map { case (pressure, (spread, _)) => (pressure, spread) }
      .filter { case (pressure, _) => pressure.toPascals <= 70000 }
      .values
      .pipe(spreads => spreads.sum / spreads.size)
  }

  def spreadAndTds(forecast: DetailedForecast): Map[Pressure, (Double, Double)] = {
    forecast.atPressure.map { case (pressure, variables) =>
      val temperature = variables.temperature.toCelsiusScale
      val exponentBase10 = 7.5 * temperature / (237.7 + temperature) // Computation of the main base 10 power exponent used below from the air temperature in °C
      val vaporPressSat = 6.11 * math.pow(10, exponentBase10) // Computation of the water vapor pressure at saturation
      // Computation of the water vapor pressure from relative humidity RH and water vapor pressure at saturation
      val vaporPress =
        if (variables.relativeHumidity > 1) (variables.relativeHumidity / 100) * vaporPressSat
        else 0.01 * vaporPressSat
      // Computation of dew point from the water vapor pressure. FIXME Why don’t we use log10 here?
      val td = (-430.22 + 237.7 * math.log(vaporPress) / (-math.log(vaporPress) + 19.08)).max(temperature)
      (pressure, (temperature - td, td))
    }
  }

  val jsonEncoder: Encoder[LocationForecasts] =
    Encoder.instance { locationForecasts =>
      Json.obj(
        "h" -> Json.fromInt(locationForecasts.elevation.toMeters.round.toInt),
        "d" -> Json.arr(
          locationForecasts.dayForecasts.map { dayForecast =>
            Json.obj(
              "th" -> Json.fromInt(dayForecast.thunderstormRisk),
              "h" -> Json.arr(
                dayForecast.hourForecasts.map { forecast =>
                  Json.obj(
                    "t" -> Json.fromString(forecast.time.format(DateTimeFormatter.ISO_DATE_TIME)),
                    "bl" -> Json.obj(
                      "h" -> Json.fromInt(forecast.boundaryLayerHeight.toMeters.round.toInt),
                      "u" -> Json.fromInt(forecast.boundaryLayerWind.u.toKilometersPerHour.round.toInt),
                      "v" -> Json.fromInt(forecast.boundaryLayerWind.v.toKilometersPerHour.round.toInt)
                    ),
                    "c" -> Json.obj(
                      "e" -> Json.fromBigDecimal(forecast.cloudCover.entire),
                      "l" -> Json.fromBigDecimal(forecast.cloudCover.low),
                      "m" -> Json.fromBigDecimal(forecast.cloudCover.middle),
                      "h" -> Json.fromBigDecimal(forecast.cloudCover.high),
                      "c" -> Json.fromBigDecimal(forecast.cloudCover.conv),
                      "b" -> Json.fromBigDecimal(forecast.cloudCover.boundary)
                    ),
                    "p" -> Json.obj(pressureLevels.map { pressure =>
                      val variables = forecast.atPressure(pressure)
                      (pressure.toPascals.round.toInt / 100).toString -> Json.obj(
                        "h" -> Json.fromInt(variables.geopotentialHeight.toMeters.round.toInt),
                        "t" -> Json.fromBigDecimal(variables.temperature.toCelsiusScale),
                        "rh" -> Json.fromBigDecimal(variables.relativeHumidity),
                        "u" -> Json.fromInt(variables.wind.u.toKilometersPerHour.round.toInt),
                        "v" -> Json.fromInt(variables.wind.v.toKilometersPerHour.round.toInt)
                      )
                    }: _*),
                    "s" -> Json.obj(
                      "t" -> Json.fromBigDecimal(forecast.surfaceTemperature.toCelsiusScale),
                      "dt" -> Json.fromBigDecimal(forecast.surfaceDewPoint.toCelsiusScale),
                      "u" -> Json.fromInt(forecast.surfaceWind.u.toKilometersPerHour.round.toInt),
                      "v" -> Json.fromInt(forecast.surfaceWind.v.toKilometersPerHour.round.toInt)
                    ),
                    "iso" -> Json.fromInt(forecast.isothermZero.toMeters.round.toInt),
                    "r" -> Json.obj(
                      "t" -> Json.fromInt(forecast.totalRain.toMillimeters.round.toInt),
                      "c" -> Json.fromInt(forecast.convectiveRain.toMillimeters.round.toInt)
                    ),
                    "mslet" -> Json.fromInt(forecast.mslet.toPascals.round.toInt / 100) // hPa
                    // TODO Irradiance, CIN, snow
                  )
                }: _*
              )
            )
          }: _*
        )
      )
    }

}

File: frontend/src/data/LocationForecasts.ts
import { ForecastMetadata } from "./ForecastMetadata";

/**
 * Forecast data for several days at a specific location.
 */
export class LocationForecasts {
  readonly elevation: number;
  readonly dayForecasts: Array<DayForecasts>;
  constructor(data: LocationForecastsData, private readonly metadata: ForecastMetadata, readonly latitude: number, readonly longitude: number) {
    this.elevation = data.h;
    this.dayForecasts = data.d.map(data => new DayForecasts(data, this.elevation));
  }

  initializationTime(): Date {
    return this.metadata.init
  }

  /** Offset (number of hours since initialization time) and date of each forecast */
  offsetAndDates(): Array<[number, Date]> {
    return this.dayForecasts
      .map(_ => _.forecasts)
      .reduce((x, y) => x.concat(y), [])
      .map(forecast => {
        return [forecast.hourOffsetSinceInitializationTime(this.metadata.init), forecast.time]
      });
  }

  /** @returns the forecast data at the given hour offset */
  atHourOffset(hourOffset: number): DetailedForecast | undefined {
    return this.dayForecasts.map(_ => _.forecasts)
      .reduce((x, y) => x.concat(y), [])
      .find(forecast => forecast.hourOffsetSinceInitializationTime(this.metadata.init) === hourOffset)
  }

}

export class DayForecasts {
  readonly thunderstormRisk: number;
  readonly forecasts: Array<DetailedForecast>;
  constructor(data: DayForecastsData, elevation: number) {
    this.thunderstormRisk = data.th;
    this.forecasts = data.h.map(data => new DetailedForecast(data, elevation));
  }
}

export class DetailedForecast {
  readonly time: Date;
  readonly xcPotential: number; // between 0 and 100
  readonly thermalVelocity: number; // m/s
  readonly boundaryLayer: DetailedBoundaryLayer;
  readonly surface: DetailedSurface;
  readonly cloudCover: number; // %
  readonly rain: DetailedRain;
  readonly meanSeaLevelPressure: number;
  readonly isothermZero: number; // m
  readonly aboveGround: Array<AboveGround>; // Sorted by ascending elevation

  constructor(data: DetailedForecastData, elevation: number) {
    this.time = new Date(data.t);
    this.xcPotential = data.xc;
    this.thermalVelocity = data.v / 10;
    this.boundaryLayer = {
      depth: data.bl.h,
      soaringLayerDepth: data.bl.c !== undefined ? data.bl.c[0] : data.bl.h,
      wind: {
        u: data.bl.u,
        v: data.bl.v
      },
      cumulusClouds: data.bl.c === undefined ? undefined : ({ bottom: data.bl.c[0] , top: data.bl.c[1] })
    };
    this.surface = {
      temperature: data.s.t,
      dewPoint: data.s.dt,
      wind: {
        u: data.s.u,
        v: data.s.v
      }
    };
    this.cloudCover = data.c / 100;
    this.rain = {
      convective: data.r.c,
      total: data.r.t
    };
    this.meanSeaLevelPressure = data.mslet;
    this.isothermZero = data.iso;

    this.aboveGround =
      data.p
        .filter(e => e.h > elevation)
        .map(entry => {
          return {
            elevation: entry.h,
            u: entry.u,
            v: entry.v,
            temperature: entry.t,
            dewPoint: entry.dt,
            cloudCover: entry.c / 100
          }
        });
  }

  hourOffsetSinceInitializationTime(initializationTime: Date): number {
    return Math.round((this.time.getTime() - initializationTime.getTime()) / 3600000)
  }

}

export type DetailedSurface = {
  temperature: number // °C
  dewPoint: number // °C
  wind: Wind
};

export type DetailedBoundaryLayer = {
  depth: number // m (AGL)
  wind: Wind
  soaringLayerDepth: number // m (AGL)
  cumulusClouds?: {
    bottom: number // m (AGL)
    top: number // m (AGL)
  }
};

export type DetailedRain = {
  convective: number // mm
  total: number // mm
};

type Wind = {
  u: number // km/h
  v: number // km/h
};

/** Various information at some elevation value */
export type AboveGround = {
  u: number // km/h
  v: number // km/h
  elevation: number // m
  temperature: number // °C
  dewPoint: number // °C
  cloudCover: number // %
};

export type LocationForecastsData = {
  h: number // elevation
  d: Array<DayForecastsData>
}

type DayForecastsData = {
  th: number // thunderstorm risk
  h: Array<DetailedForecastData>
}

export type DetailedForecastData = {
  t: string // Forecast time
  xc: number // XC Potential, between 0 and 100
  // Boundary layer
  bl: {
    // Depth (m AGL)
    h: number,
    // Wind
    u: number,
    v: number,
    // Cumulus clouds bottom and top (m AGL)
    c?: [number, number]
  }
  v: number, // Thermal velocity
  // Above ground variables
  p: Array<{
    h: number, // altitude
    t: number, // temperature
    dt: number, // dew point temperature
    // wind
    u: number,
    v: number,
    c: number // cloud cover
  }>,
  // Surface
  s: {
    t: number, // temperature
    dt: number, // dew point temperature
    // Wind
    u: number,
    v: number
  }
  // Isotherm 0°C
  iso: number,
  // Rain
  r: {
    t: number, // total
    c: number // convective
  },
  // Mean sea level pressure 
  mslet: number, // hPa
  c: number // Between 0 and 100
}

export const modelResolution = 25 // Hundredths of degrees

/**
 * Return the closest point of the underlying model grid, in hundreths of degrees.
 * 
 * E.g., for GFS `normalizeCoordinates(46.1234, 7.5678) == [4600, 750]`.
 */
export const normalizeCoordinates =
  (latitude: number, longitude: number): [number, number] => {
    return [
      Math.floor(((latitude * 100) + modelResolution / 2) / modelResolution) * modelResolution,
      Math.floor(((longitude * 100) + modelResolution / 2) / modelResolution) * modelResolution
    ]
  }
